File: 3Sum.cpp
class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < n; ++j) {
                if (j - 1 > i && nums[j] == nums[j - 1]) continue;
                int c = -nums[i] - nums[j];
                int idx = lower_bound(nums.begin() + j + 1, nums.end(), c) - nums.begin();
                if (idx >= j + 1 && idx < n && nums[idx] == c) {
                    vector<int> temp = {nums[i], nums[j], nums[idx]};
                    ans.push_back(temp);
                }
            }
        }
        return ans;
    }
};

File: 3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        nums_dict = collections.defaultdict(list)
        for idx, i in enumerate(nums):
                nums_dict[i].append(idx)    
        ans = []
        prev1 = 0
        for i in range(len(nums)):
            if i > 0 and nums[i] == prev1:
                continue
            prev1 = nums[i]
            for j in range(i+1, len(nums)):
                target = -nums[i]-nums[j]
                if target in nums_dict:
                    for k in nums_dict[target]:
                        if k != i and k != j:
                            ans.append(tuple(sorted([nums[i], nums[j], -nums[i]-nums[j]])))
                            break
        ans = list(set(ans))
        for idx in range(len(ans)):
            ans[idx] = list(ans[idx])
        return ans